api/src/main/java/net/countercraft/movecraft/craft/datatag/CraftDataTagContainer.java

```diff
@@ -0,0 +1,53 @@
+package net.countercraft.movecraft.craft.datatag;
+
+import net.countercraft.movecraft.craft.Craft;
+import org.bukkit.NamespacedKey;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class CraftDataTagContainer extends HashMap<CraftDataTagKey<?>, Object> {
+
+    public static final Map<NamespacedKey, CraftDataTagKey<?>> REGISTERED_TAGS = new HashMap<>();
+
+    public static <T extends ICraftDataTag> CraftDataTagKey<T> tryRegisterTagKey(final NamespacedKey key, final Function<Craft, T> supplier) {
+        if (REGISTERED_TAGS.containsKey(key)) {
+            throw new IllegalArgumentException("Duplicate keys are not allowed!");
```

Comment:
Rather than be `tryRegisterTagKey` we could rename this to be `registerTagKey` and annotate it with `throws IllegalArgumentException`.

Reply:
Ye, that would be cleaner, also doesn't lead to unexpected exceptions